### ISSUE TITLE
Errors "-Wsign-conversion"

### ISSUE THREAD

**Issue by wosrediinanatour** [warmcat/libwebsockets#2173]
Commit f64de6386cfeec6848b1117dd8e351bde7344e5b has following compile errors, when performing
```
cmake \
    -D CMAKE_BUILD_TYPE=DEBUG \
    -D LWS_WITH_LIBUV=ON \
    -D LWS_WITH_STATIC=ON \
    -D LWS_WITHOUT_BUILTIN_SHA1=ON \
    -D LWS_WITHOUT_TESTAPPS=ON \
    -D LWS_WITHOUT_TEST_SERVER=ON \
    -D LWS_WITHOUT_TEST_SERVER_EXTPOLL=ON \
    -D LWS_WITHOUT_TEST_PING=ON \
    -D LWS_WITHOUT_TEST_ECHO=ON \
    -D LWS_WITHOUT_TEST_CLIENT=OFF \
    -D LWS_WITHOUT_TEST_FRAGGLE=ON \
    -D LWS_USE_BUNDLED_ZLIB=OFF \
    -D LWS_HAVE_UV_VERSION=OFF \
    -D LWS_HAVE_NEW_UV_VERSION_H=ON \
    -D LWS_WITHOUT_EXTAENSIONS=0 \
    ..
```
with gcc (GCC) 9.1.1 20190605 (Red Hat 9.1.1-2):

```
cd /root/rpmbuild/BUILD/libwebsockets/build/lib && /opt/rh/devtoolset-9/root/usr/bin/cc  -DLWS_BUILDING_STATIC -Wall -Wconversion -Wsign-compare -Wstrict-aliasing -fvisibility=hidden -Wundef  -Wuninitialized -Wtype-limits -Wignored-quali
fiers -Wno-error=sign-conversion  -Werror -g -I/root/rpmbuild/BUILD/libwebsockets/include -I/root/rpmbuild/BUILD/libwebsockets/build -I/root/rpmbuild/BUILD/libwebsockets/lib -I/root/rpmbuild/BUILD/libwebsockets/lib/. -I/root/rpmbuild/BUI
LD/libwebsockets/lib/plat/unix/. -I/root/rpmbuild/BUILD/libwebsockets/lib/tls/. -I/root/rpmbuild/BUILD/libwebsockets/lib/core/. -I/root/rpmbuild/BUILD/libwebsockets/lib/misc/. -I/root/rpmbuild/BUILD/libwebsockets/lib/system/./async-dns -
I/root/rpmbuild/BUILD/libwebsockets/lib/system/smd/. -I/root/rpmbuild/BUILD/libwebsockets/lib/system/metrics/. -I/root/rpmbuild/BUILD/libwebsockets/lib/core-net/. -I/root/rpmbuild/BUILD/libwebsockets/lib/roles/. -I/root/rpmbuild/BUILD/li
bwebsockets/lib/roles/http/. -I/root/rpmbuild/BUILD/libwebsockets/lib/roles/http/./compression -I/root/rpmbuild/BUILD/libwebsockets/lib/roles/h1/. -I/root/rpmbuild/BUILD/libwebsockets/lib/roles/h2/. -I/root/rpmbuild/BUILD/libwebsockets/l
ib/roles/ws/. -I/root/rpmbuild/BUILD/libwebsockets/lib/roles/raw-skt/. -I/root/rpmbuild/BUILD/libwebsockets/lib/roles/raw-file/. -I/root/rpmbuild/BUILD/libwebsockets/lib/roles/listen/. -I/root/rpmbuild/BUILD/libwebsockets/lib/event-libs/
. -I/root/rpmbuild/BUILD/libwebsockets/lib/event-libs/poll/../poll -I/root/rpmbuild/BUILD/libwebsockets/lib/event-libs/libuv/.    -o CMakeFiles/websockets.dir/system/system.c.o   -c /root/rpmbuild/BUILD/libwebsockets/lib/system/system.c
In file included from /root/rpmbuild/BUILD/libwebsockets/include/libwebsockets.h:288,
                 from /root/rpmbuild/BUILD/libwebsockets/lib/core/./private-lib-core.h:140,
                 from /root/rpmbuild/BUILD/libwebsockets/lib/tls/openssl/openssl-server.c:25:
/root/rpmbuild/BUILD/libwebsockets/lib/tls/openssl/openssl-server.c: In function 'lws_tls_server_vhost_backend_init':
/root/rpmbuild/BUILD/libwebsockets/lib/tls/openssl/openssl-server.c:584:5: warning: conversion to 'long int' from 'long unsigned int' may change the sign of the result [-Wsign-conversion]
  584 |     (unsigned long)
      |     ^
/root/rpmbuild/BUILD/libwebsockets/lib/tls/openssl/openssl-server.c:596:11: warning: conversion to 'long int' from 'long unsigned int' may change the sign of the result [-Wsign-conversion]
  596 |           (unsigned long)
      |           ^
`
``

```cd /root/rpmbuild/BUILD/libwebsockets/build/lib && /opt/rh/devtoolset-9/root/usr/bin/cc  -DLWS_BUILDING_SHARED -Dwebsockets_shared_EXPORTS -Wall -Wconversion -Wsign-compare -Wstrict-aliasing -fvisibility=hidden -Wundef  -Wuninitialized -
Wtype-limits -Wignored-qualifiers -Wno-error=sign-conversion  -Werror -g -fPIC -I/root/rpmbuild/BUILD/libwebsockets/include -I/root/rpmbuild/BUILD/libwebsockets/build -I/root/rpmbuild/BUILD/libwebsockets/lib -I/root/rpmbuild/BUILD/libweb
sockets/lib/. -I/root/rpmbuild/BUILD/libwebsockets/lib/plat/unix/. -I/root/rpmbuild/BUILD/libwebsockets/lib/tls/. -I/root/rpmbuild/BUILD/libwebsockets/lib/core/. -I/root/rpmbuild/BUILD/libwebsockets/lib/misc/. -I/root/rpmbuild/BUILD/libw
ebsockets/lib/system/./async-dns -I/root/rpmbuild/BUILD/libwebsockets/lib/system/smd/. -I/root/rpmbuild/BUILD/libwebsockets/lib/system/metrics/. -I/root/rpmbuild/BUILD/libwebsockets/lib/core-net/. -I/root/rpmbuild/BUILD/libwebsockets/lib
/roles/. -I/root/rpmbuild/BUILD/libwebsockets/lib/roles/http/. -I/root/rpmbuild/BUILD/libwebsockets/lib/roles/http/./compression -I/root/rpmbuild/BUILD/libwebsockets/lib/roles/h1/. -I/root/rpmbuild/BUILD/libwebsockets/lib/roles/h2/. -I/r
oot/rpmbuild/BUILD/libwebsockets/lib/roles/ws/. -I/root/rpmbuild/BUILD/libwebsockets/lib/roles/raw-skt/. -I/root/rpmbuild/BUILD/libwebsockets/lib/roles/raw-file/. -I/root/rpmbuild/BUILD/libwebsockets/lib/roles/listen/. -I/root/rpmbuild/B
UILD/libwebsockets/lib/event-libs/. -I/root/rpmbuild/BUILD/libwebsockets/lib/event-libs/poll/../poll -I/root/rpmbuild/BUILD/libwebsockets/lib/event-libs/libuv/.    -o CMakeFiles/websockets_shared.dir/core/alloc.c.o   -c /root/rpmbuild/BU
ILD/libwebsockets/lib/core/alloc.c
/usr/bin/cmake -E cmake_progress_report /root/rpmbuild/BUILD/libwebsockets/build/CMakeFiles 26
[ 39%] In file included from /root/rpmbuild/BUILD/libwebsockets/include/libwebsockets.h:288,
                 from /root/rpmbuild/BUILD/libwebsockets/lib/core/./private-lib-core.h:140,
                 from /root/rpmbuild/BUILD/libwebsockets/lib/tls/openssl/openssl-client.c:32:
/root/rpmbuild/BUILD/libwebsockets/lib/tls/openssl/openssl-client.c: In function 'lws_tls_client_create_vhost_context':
/root/rpmbuild/BUILD/libwebsockets/lib/tls/openssl/openssl-client.c:778:9: warning: conversion to 'long int' from 'long unsigned int' may change the sign of the result [-Wsign-conversion]
  778 |         (unsigned long)
      |         ^
/root/rpmbuild/BUILD/libwebsockets/lib/tls/openssl/openssl-client.c:790:11: warning: conversion to 'long int' from 'long unsigned int' may change the sign of the result [-Wsign-conversion]
  790 |           (unsigned long)
      |           ^

```

```cd /root/rpmbuild/BUILD/libwebsockets/build/lib && /opt/rh/devtoolset-9/root/usr/bin/cc  -DLWS_BUILDING_SHARED -Dwebsockets_shared_EXPORTS -Wall -Wconversion -Wsign-compare -Wstrict-aliasing -fvisibility=hidden -Wundef  -Wuninitialized -
Wtype-limits -Wignored-qualifiers -Wno-error=sign-conversion  -Werror -g -fPIC -I/root/rpmbuild/BUILD/libwebsockets/include -I/root/rpmbuild/BUILD/libwebsockets/build -I/root/rpmbuild/BUILD/libwebsockets/lib -I/root/rpmbuild/BUILD/libweb
sockets/lib/. -I/root/rpmbuild/BUILD/libwebsockets/lib/plat/unix/. -I/root/rpmbuild/BUILD/libwebsockets/lib/tls/. -I/root/rpmbuild/BUILD/libwebsockets/lib/core/. -I/root/rpmbuild/BUILD/libwebsockets/lib/misc/. -I/root/rpmbuild/BUILD/libw
ebsockets/lib/system/./async-dns -I/root/rpmbuild/BUILD/libwebsockets/lib/system/smd/. -I/root/rpmbuild/BUILD/libwebsockets/lib/system/metrics/. -I/root/rpmbuild/BUILD/libwebsockets/lib/core-net/. -I/root/rpmbuild/BUILD/libwebsockets/lib
/roles/. -I/root/rpmbuild/BUILD/libwebsockets/lib/roles/http/. -I/root/rpmbuild/BUILD/libwebsockets/lib/roles/http/./compression -I/root/rpmbuild/BUILD/libwebsockets/lib/roles/h1/. -I/root/rpmbuild/BUILD/libwebsockets/lib/roles/h2/. -I/r
oot/rpmbuild/BUILD/libwebsockets/lib/roles/ws/. -I/root/rpmbuild/BUILD/libwebsockets/lib/roles/raw-skt/. -I/root/rpmbuild/BUILD/libwebsockets/lib/roles/raw-file/. -I/root/rpmbuild/BUILD/libwebsockets/lib/roles/listen/. -I/root/rpmbuild/B
UILD/libwebsockets/lib/event-libs/. -I/root/rpmbuild/BUILD/libwebsockets/lib/event-libs/poll/../poll -I/root/rpmbuild/BUILD/libwebsockets/lib/event-libs/libuv/.    -o CMakeFiles/websockets_shared.dir/roles/netlink/ops-netlink.c.o   -c /r
oot/rpmbuild/BUILD/libwebsockets/lib/roles/netlink/ops-netlink.c
/root/rpmbuild/BUILD/libwebsockets/lib/roles/netlink/ops-netlink.c: In function 'rops_handle_POLLIN_netlink':
/root/rpmbuild/BUILD/libwebsockets/lib/roles/netlink/ops-netlink.c:167:36: warning: conversion to 'unsigned int' from 'int' may change the sign of the result [-Wsign-conversion]
  167 |   for ( ; RTA_OK(ra, ra_len); ra = RTA_NEXT(ra, ra_len)) {
      |                                    ^~~~~~~~
/root/rpmbuild/BUILD/libwebsockets/lib/roles/netlink/ops-netlink.c:170:39: warning: unsigned conversion from 'int' to 'long unsigned int' changes value from '-4' to '18446744073709551612' [-Wsign-conversion]
  170 |     lws_sa46_copy_address(&robj.dest, RTA_DATA(ra),
      |                                       ^~~~~~~~
/root/rpmbuild/BUILD/libwebsockets/lib/roles/netlink/ops-netlink.c:176:13: warning: unsigned conversion from 'int' to 'long unsigned int' changes value from '-4' to '18446744073709551612' [-Wsign-conversion]
  176 |             RTA_DATA(ra),
      |             ^~~~~~~~
/root/rpmbuild/BUILD/libwebsockets/lib/roles/netlink/ops-netlink.c:183:27: warning: unsigned conversion from 'int' to 'long unsigned int' changes value from '-4' to '18446744073709551612' [-Wsign-conversion]
  183 |     robj.if_idx = *(char*)RTA_DATA(ra);
      |                           ^~~~~~~~
/root/rpmbuild/BUILD/libwebsockets/lib/roles/netlink/ops-netlink.c:186:9: warning: unsigned conversion from 'int' to 'long unsigned int' changes value from '-4' to '18446744073709551612' [-Wsign-conversion]
  186 |     p = RTA_DATA(ra);
      |         ^~~~~~~~
```

**Comment by lws-team** [comment]
> -D LWS_WITHOUT_EXTAENSIONS=0 \

There's a typo there FYI, your probably wanted ..._EXTENSIONS

```
/root/rpmbuild/BUILD/libwebsockets/lib/tls/openssl/openssl-client.c:778:9: warning: conversion to 'long int' from 'long unsigned int' may change the sign of the result [-Wsign-conversion]
  778 |         (unsigned long)
```

Openssl changed the signature for `SSL_CTX_clear_/set_options()` from long to unsigned long at some point after 1.0.2

https://www.openssl.org/docs/man1.0.2/man3/SSL_CTX_clear_options.html

```
$ grep SSL_CTX_set_options /usr/include -R
/usr/include/openssl/ssl.h:unsigned long SSL_CTX_set_options(SSL_CTX *ctx, unsigned long op);
$ rpm -q openssl-devel
openssl-devel-1.1.1i-1.fc33.x86_64
```

all the platforms in Sai are OK with unsigned long... what's your openssl version?

> /root/rpmbuild/BUILD/libwebsockets/lib/roles/netlink/ops-netlink.c:167:36: warning: conversion to 'unsigned int' from 'int' may change the sign of the result [-Wsign-conversion]
>  167 |   for ( ; RTA_OK(ra, ra_len); ra = RTA_NEXT(ra, ra_len)) {

This is related to your kernel headers... again all 23 platforms in Sai can build this without warnings / errors... what's the age of your kernel headers?  For Fedora 33 kernel-headers-5.9.13-200.fc33.x86_64 the related bits in /usr/include/linux/rtnetlink.h look like this

```
#define RTA_ALIGNTO     4U
#define RTA_ALIGN(len) ( ((len)+RTA_ALIGNTO-1) & ~(RTA_ALIGNTO-1) )
#define RTA_OK(rta,len) ((len) >= (int)sizeof(struct rtattr) && \
                         (rta)->rta_len >= sizeof(struct rtattr) && \
                         (rta)->rta_len <= (len))
#define RTA_NEXT(rta,attrlen)   ((attrlen) -= RTA_ALIGN((rta)->rta_len), \
                                 (struct rtattr*)(((char*)(rta)) + RTA_ALIGN((rta)->rta_len)))
```

**Comment by wosrediinanatour** [comment]
Thanks for the typo.

You are right - the openssl and the kernel headers are old.

The packages are:
```
openssl-devel-1.0.2k-19.el7.x86_64
```
with
```
# define SSL_CTX_set_options(ctx,op) \
        SSL_CTX_ctrl((ctx),SSL_CTRL_OPTIONS,(op),NULL)
...
long SSL_CTX_ctrl(SSL_CTX *ctx, int cmd, long larg, void *parg);
```
It is 
```
kernel-headers-3.10.0-1127.el7.x86_64 
```
with
```
#define RTA_ALIGNTO     4
#define RTA_ALIGN(len) ( ((len)+RTA_ALIGNTO-1) & ~(RTA_ALIGNTO-1) )
#define RTA_OK(rta,len) ((len) >= (int)sizeof(struct rtattr) && \
                         (rta)->rta_len >= sizeof(struct rtattr) && \
                         (rta)->rta_len <= (len))
#define RTA_NEXT(rta,attrlen)   ((attrlen) -= RTA_ALIGN((rta)->rta_len), \
                                 (struct rtattr*)(((char*)(rta)) + RTA_ALIGN((rta)->rta_len)))
```

**Comment by lws-team** [comment]
I can't see the difference in the kernel headers that makes the complaint... as I say there's no problem on any of the Sai platforms which includes, eg, Centos 8 / Stream and that's on gcc 8.4.1, he has openssl 1.1.1.

For the openssl, the change is not documented in the version of openssl it happened in, unlike a lot of their changes where they mention it in the manpages.  And I don't know a good, clean way to detect the change at cmake.

Can you upgrade to Centos 8 / Stream and more modern pieces?

**Comment by wosrediinanatour** [comment]
I guess it is the 4u vs. 4 in the kernel header.

It is not possible to update. Most probably, the errors should not harm. What do you thing?

**Comment by lws-team** [comment]
I missed the u... yes that will be it.

I added CentOS 7 to Sai after a lot of struggle, he can build all the scenarios except mbedtls, there's no mbedtls package on CentOS7.  There's no libuv either, I built it and also updated the cmake to HEAD since although the packaged one can built lws and sai fine, it's too old to have all the ctest options we use.

I pushed a patch on main that works around the problems you mentioned here and a bunch more that came out from building the other scenarios.

https://libwebsockets.org/git/libwebsockets/commit?id=d4a21db2bc424cf820c0f1d8573c618f1ebc6369
